networkx/generators/geometric.py

```diff
@@ -844,3 +845,205 @@ def thresholded_random_geometric_graph(
+    and the exponent of the powerlaw distribution of the hidden degrees is 2.7
+    with mean value of 5.
+
+    Generate a network with predefined hidden degrees:
+
+    >>> kappas = {i: 10 for i in range(100)}
+    >>> G = nx.geometric_soft_configuration_graph(beta=2.5, kappas=kappas)
+
+    Create a $\mathbb{S}^1$ model with 100 nodes. The $\beta$ parameter is set to 2.5
+    and all nodes with hidden degree $\kappa=10$.
+
+
+    References
+    ----------
+    .. [1] Serrano, M. Ángeles, Dmitri, Krioukov, and Marián, Boguñá. "Self-Similarity
+    of Complex Networks and Hidden Metric Spaces". Phys. Rev. Lett. 100 (2008): 078701.
+
+    .. [2] van der Kolk, Jasper, M. Ángeles Serrano, and Marián Boguñá. "An anomalous
+    topological phase transition in spatial random graphs."
+    Communications Physics 5.1 (2022): 245.
+
+    .. [3] Krioukov, Dmitri, Fragkiskos Papadopoulos, Maksim Kitsak, Amin Vahdat,
+    and Marián Boguná. "Hyperbolic geometry of complex networks."
+    Physical Review E 82, no. 3 (2010): 036106.
+
+    """
+    if beta <= 0:
+        raise nx.NetworkXError("The parameter beta cannot be smaller or equal to 0.")
+
+    if kappas is not None:
+        if not all((n is None, gamma is None, mean_degree is None)):
+            raise nx.NetworkXError(
+                "When kappas is input, n, gamma and mean_degree must not be."
+            )
+
+        n = len(kappas)
+        mean_degree = sum(kappas) / len(kappas)
+    else:
+        if any((n is None, gamma is None, mean_degree is None)):
+            raise nx.NetworkXError(
+                "Please provide either kappas, or all 3 parameters: n, gamma and mean_degree."
+            )
+
+        # Generate hidden degrees from the powerlaw distribution with given exponent `gamma`
+        #  and mean value `mean_degree`
+        kappa_0 = (
+            (1 - 1 / n)
+            / (1 - n ** ((2 - gamma) / (gamma - 1)))
+            * (gamma - 2)
+            / (gamma - 1)
+            * mean_degree
+        )
+        kappa_c = kappa_0 * n ** (1 / (gamma - 1))
+
+        kappas = {}
+        for i in range(n):
+            kappas[i] = kappa_0 * (
+                1 - seed.uniform(0, 1) * (1 - (kappa_c / kappa_0) ** (1 - gamma))
+            ) ** (1 / (1 - gamma))
+
+    G = nx.empty_graph(kappas.keys())
+    R = n / (2 * math.pi)
+
+    # Approximate values for mu in the thermodynamic limit (when n -> infinity)
+    if beta > 1:
+        mu = beta * math.sin(math.pi / beta) / (2 * math.pi * mean_degree)
+    elif beta == 1:
+        mu = 1 / (2 * mean_degree * math.log(n))
+    else:
+        mu = (1 - beta) / (2**beta * mean_degree * n ** (1 - beta))
+
+    # Generate random positions on a circle
+    thetas = {k: seed.uniform(0, 2 * math.pi) for k in kappas}
+    mapping = dict(zip(range(n), kappas.keys()))
+
+    for i in range(n):
+        for j in range(i):
+            k, l = mapping[i], mapping[j]
```

Comment:
It's traditional (at least in networkx) to denote pairs of nodes using `u` and `v` rather than `i, j, k, l`. Also, we don't need to convert back and forth between integers and nodes. What do you think of:
```
    G = nx.Graph()
    for u in kappas:
        for v in list(G):
            <compute p_ij using u and v instead of k and l>
            if seed.random() < p_ij:
                G.add_edge(u, v)
        G.add_node(u)  # ensures u becomes a node in G
```
We need the `list(G)` because we may add the node u during this loop.
And since no edges might be added, we put the `add_node` after the loop over `v`.

**silliness:**  Another possible convenience is that $\tau$ is defined to be $2*\pi$. And python has `math.tau` in addition to `math.pi`. I guess there is a long-standing debate about whether $\tau$ or $\pi$ should be the dominant circle constant taught in school.  So, if you want to expand the reputation of `tau` you can use `math.tau` wherever you have `2 * math.pi`.